File: packages/js/src/normalise.js
'use strict';

const dictionary = require('./dictionary.json');

const VOWELS = new Set(['a', 'e', 'i', 'o', 'u']);
const DIGRAPHS = ['ng', 'ny', 'sy', 'kh', 'gh'];
const DIPHTHONGS = ['ai', 'au', 'oi'];

function titleCase(word) {
  if (!word) return word;
  return word[0].toUpperCase() + word.slice(1).toLowerCase();
}

// Lightweight Malay syllabifier. Used for tokens not in the dictionary so the
// tts_ready output is consistent across the whole address.
function syllabify(word) {
  if (word.length <= 2) return word;
  const lower = word.toLowerCase();

  // Tokenise the word into atomic units, keeping digraphs (ng, ny, ...) together.
  const units = [];
  for (let i = 0; i < word.length; ) {
    const two = lower.slice(i, i + 2);
    if (DIGRAPHS.includes(two)) {
      units.push({ chars: word.slice(i, i + 2), type: 'consonant' });
      i += 2;
    } else {
      const ch = lower[i];
      units.push({ chars: word[i], type: VOWELS.has(ch) ? 'vowel' : 'consonant' });
      i += 1;
    }
  }

  // Collapse word-final diphthongs (ai, au, oi) into a single vowel unit so
  // names like "Sungai" syllabify as "Su-ngai", not "Su-nga-i".
  if (units.length >= 2) {
    const last = units[units.length - 1];
    const prev = units[units.length - 2];
    if (last.type === 'vowel' && prev.type === 'vowel') {
      const pair = (prev.chars + last.chars).toLowerCase();
      if (DIPHTHONGS.includes(pair)) {
        units.splice(units.length - 2, 2, { chars: prev.chars + last.chars, type: 'vowel' });
      }
    }
  }

  const vowelIdx = [];
  units.forEach((u, j) => { if (u.type === 'vowel') vowelIdx.push(j); });
  if (vowelIdx.length <= 1) return word;

  // Maximum-onset rule: a lone consonant between vowels starts the next
  // syllable (V-CV); two or more split one-to-the-left (VC-CV / VC-CCV).
  const boundaries = [0];
  for (let k = 0; k < vowelIdx.length - 1; k++) {
    const v1 = vowelIdx[k];
    const v2 = vowelIdx[k + 1];
    const between = v2 - v1 - 1;
    if (between === 0)      boundaries.push(v2);
    else if (between === 1) boundaries.push(v1 + 1);
    else                    boundaries.push(v1 + 2);
  }

  const syllables = [];
  for (let b = 0; b < boundaries.length; b++) {
    const start = boundaries[b];
    const end = b + 1 < boundaries.length ? boundaries[b + 1] : units.length;
    syllables.push(units.slice(start, end).map(u => u.chars).join(''));
  }
  return syllables.join('-');
}

function isAllCaps(s) {
  return /[A-Z]/.test(s) && s === s.toUpperCase();
}

function isWhitespace(t) { return /^\s+$/.test(t); }
function isPunctuation(t) { return /^[,.;:()\/]+$/.test(t); }

// A bare number or alphanumeric code that starts with a digit (1, 12, 1A, 2B).
// Used by the numeric_suffix_is_code rule.
function isCodeToken(t) { return /^\d[\dA-Za-z]*$/.test(t); }

function nextContentToken(tokens, fromIndex) {
  for (let i = fromIndex + 1; i < tokens.length; i++) {
    const t = tokens[i];
    if (!isWhitespace(t) && !isPunctuation(t)) return t;
  }
  return null;
}

function normalise(input) {
  if (typeof input !== 'string') {
    throw new TypeError('input must be a string');
  }

  // Split on whitespace and a small set of address punctuation, keeping the
  // separators so we can reassemble the original spacing.
  const tokens = input.split(/(\s+|[,.;:()\/])/).filter(t => t !== '');

  const normalisedParts = [];
  const ttsParts = [];
  const expansionsApplied = [];
  const ambiguousFlags = [];
  let previousContentToken = null;

  for (let idx = 0; idx < tokens.length; idx++) {
    const token = tokens[idx];
    if (isWhitespace(token) || isPunctuation(token)) {
      normalisedParts.push(token);
      ttsParts.push(token);
      continue;
    }

    // Detect a number glued to letters, e.g. "5KG", so we can both look up the
    // alpha part in the dictionary and apply the weight heuristic.
    const glued = token.match(/^(\d+)([A-Za-z]+)$/);
    const numericAttached = glued ? glued[1] : null;
    const alphaPart = glued ? glued[2] : token;
    const upper = alphaPart.toUpperCase();
    const entry = dictionary[upper];

    if (entry) {
      // Weight heuristic: a numeric prefix (in-token or as the previous token)
      // means we're looking at a unit like "5 KG of rice", not "Kampung".
      if (entry.rule === 'numeric_prefix_is_weight') {
        const numericContext = numericAttached !== null
          || (previousContentToken !== null && /^\d+$/.test(previousContentToken));
        if (numericContext) {
          normalisedParts.push(token);
          ttsParts.push(token);
          ambiguousFlags.push({
            token,
            chosen: token,
            alternative: entry.expansion,
            reason: `numeric context suggests ${entry.ambiguous_with}; left unexpanded`,
          });
          previousContentToken = token;
          continue;
        }
      }

      // Section-code heuristic: when the following token starts with a digit
      // (e.g. "KT 1", "KT 2B"), treat KT as a section prefix and leave it
      // alone. With no numeric follower, fall through to the expansion.
      if (entry.rule === 'numeric_suffix_is_code') {
        const next = nextContentToken(tokens, idx);
        if (next && isCodeToken(next)) {
          normalisedParts.push(token);
          ttsParts.push(token);
          ambiguousFlags.push({
            token,
            chosen: token,
            alternative: entry.expansion,
            reason: `following token "${next}" suggests ${entry.ambiguous_with}; left unexpanded`,
          });
          previousContentToken = token;
          continue;
        }
      }

      const expansion = entry.expansion;
      const syllables = entry.syllables;
      const display = numericAttached ? `${numericAttached} ${expansion}` : expansion;
      const tts = numericAttached ? `${numericAttached} ${syllables}` : syllables;

      normalisedParts.push(display);
      ttsParts.push(tts);
      expansionsApplied.push({ from: token, to: expansion });

      // Only flag when we *defaulted* — i.e. picked one of multiple plausible
      // meanings without a deterministic rule to resolve it. KG resolved by
      // numeric_prefix_is_weight is not a guess; SG defaulting to Sungai is.
      if (entry.ambiguous_with && entry.rule === 'default_to_malay') {
        ambiguousFlags.push({
          token,
          chosen: expansion,
          alternative: entry.ambiguous_with,
          reason: `defaulted to ${expansion}; could also mean ${entry.ambiguous_with}`,
        });
      }
    } else {
      // Title-case only purely alphabetic all-caps tokens. Mixed alphanumeric
      // tokens (KT1, A12, 50450) are codes — preserve them verbatim.
      const pureAlpha = /^[A-Za-z]+$/.test(alphaPart);
      const display = (pureAlpha && isAllCaps(alphaPart)) ? titleCase(alphaPart) : alphaPart;
      const tts = syllabify(display);
      normalisedParts.push(numericAttached ? numericAttached + display : display);
      ttsParts.push(numericAttached ? numericAttached + tts : tts);
    }

    previousContentToken = token;
  }

  return {
    normalised: normalisedParts.join(''),
    tts_ready: ttsParts.join(''),
    expansions_applied: expansionsApplied,
    ambiguous_flags: ambiguousFlags,
  };
}

module.exports = { normalise, syllabify };

File: packages/python/bm_address/normalise.py
"""Address normaliser. Mirrors packages/js/src/normalise.js."""

from __future__ import annotations

import re
from typing import Any

from .dictionary import DICTIONARY

VOWELS = set("aeiou")
DIGRAPHS = ("ng", "ny", "sy", "kh", "gh")
DIPHTHONGS = ("ai", "au", "oi")

_TOKEN_RE = re.compile(r"(\s+|[,.;:()/])")
_GLUED_RE = re.compile(r"^(\d+)([A-Za-z]+)$")
_WHITESPACE_RE = re.compile(r"^\s+$")
_PUNCT_RE = re.compile(r"^[,.;:()/]+$")
_NUMERIC_RE = re.compile(r"^\d+$")
# A bare number or alphanumeric code that starts with a digit (1, 12, 1A, 2B).
# Used by the numeric_suffix_is_code rule.
_CODE_RE = re.compile(r"^\d[\dA-Za-z]*$")
_PURE_ALPHA_RE = re.compile(r"^[A-Za-z]+$")


def _title_case(word: str) -> str:
    if not word:
        return word
    return word[0].upper() + word[1:].lower()


def _is_all_caps(s: str) -> bool:
    return any(c.isalpha() for c in s) and s == s.upper()


def _next_content_token(tokens: list[str], from_index: int) -> str | None:
    for i in range(from_index + 1, len(tokens)):
        t = tokens[i]
        if not _WHITESPACE_RE.match(t) and not _PUNCT_RE.match(t):
            return t
    return None


def syllabify(word: str) -> str:
    """Split a Malay word into hyphenated syllables.

    Used for words not in the dictionary so tts_ready stays consistent across
    the whole address. Dictionary entries supply their own syllabification.
    """
    if len(word) <= 2:
        return word
    lower = word.lower()

    units = []
    i = 0
    while i < len(word):
        two = lower[i:i + 2]
        if two in DIGRAPHS:
            units.append({"chars": word[i:i + 2], "type": "consonant"})
            i += 2
        else:
            ch = lower[i]
            t = "vowel" if ch in VOWELS else "consonant"
            units.append({"chars": word[i], "type": t})
            i += 1

    # Collapse word-final diphthongs so "Sungai" syllabifies "Su-ngai", not
    # "Su-nga-i".
    if len(units) >= 2:
        last, prev = units[-1], units[-2]
        if last["type"] == "vowel" and prev["type"] == "vowel":
            pair = (prev["chars"] + last["chars"]).lower()
            if pair in DIPHTHONGS:
                units = units[:-2] + [{"chars": prev["chars"] + last["chars"], "type": "vowel"}]

    vowel_idx = [j for j, u in enumerate(units) if u["type"] == "vowel"]
    if len(vowel_idx) <= 1:
        return word

    # Maximum-onset rule: one consonant between vowels starts the next syllable
    # (V-CV); two or more split one-to-the-left (VC-CV / VC-CCV).
    boundaries = [0]
    for k in range(len(vowel_idx) - 1):
        v1, v2 = vowel_idx[k], vowel_idx[k + 1]
        between = v2 - v1 - 1
        if between == 0:
            boundaries.append(v2)
        elif between == 1:
            boundaries.append(v1 + 1)
        else:
            boundaries.append(v1 + 2)

    syllables = []
    for b, start in enumerate(boundaries):
        end = boundaries[b + 1] if b + 1 < len(boundaries) else len(units)
        syllables.append("".join(u["chars"] for u in units[start:end]))
    return "-".join(syllables)


def normalise(input_str: str) -> dict[str, Any]:
    """Expand an abbreviated Malaysian address.

    Returns a dict with: normalised, tts_ready, expansions_applied,
    ambiguous_flags.
    """
    if not isinstance(input_str, str):
        raise TypeError("input must be a string")

    tokens = [t for t in _TOKEN_RE.split(input_str) if t != ""]

    normalised_parts: list[str] = []
    tts_parts: list[str] = []
    expansions_applied: list[dict[str, str]] = []
    ambiguous_flags: list[dict[str, str]] = []
    previous_content_token: str | None = None

    for idx, token in enumerate(tokens):
        if _WHITESPACE_RE.match(token) or _PUNCT_RE.match(token):
            normalised_parts.append(token)
            tts_parts.append(token)
            continue

        glued = _GLUED_RE.match(token)
        numeric_attached = glued.group(1) if glued else None
        alpha_part = glued.group(2) if glued else token
        upper = alpha_part.upper()
        entry = DICTIONARY.get(upper)

        if entry is not None:
            if entry.get("rule") == "numeric_prefix_is_weight":
                numeric_context = numeric_attached is not None or (
                    previous_content_token is not None
                    and _NUMERIC_RE.match(previous_content_token) is not None
                )
                if numeric_context:
                    normalised_parts.append(token)
                    tts_parts.append(token)
                    ambiguous_flags.append({
                        "token": token,
                        "chosen": token,
                        "alternative": entry["expansion"],
                        "reason": f"numeric context suggests {entry['ambiguous_with']}; left unexpanded",
                    })
                    previous_content_token = token
                    continue

            # Section-code heuristic: when the following token starts with a
            # digit (e.g. "KT 1", "KT 2B"), treat KT as a section prefix and
            # leave it alone. With no numeric follower, fall through to expand.
            if entry.get("rule") == "numeric_suffix_is_code":
                nxt = _next_content_token(tokens, idx)
                if nxt is not None and _CODE_RE.match(nxt) is not None:
                    normalised_parts.append(token)
                    tts_parts.append(token)
                    ambiguous_flags.append({
                        "token": token,
                        "chosen": token,
                        "alternative": entry["expansion"],
                        "reason": f'following token "{nxt}" suggests {entry["ambiguous_with"]}; left unexpanded',
                    })
                    previous_content_token = token
                    continue

            expansion = entry["expansion"]
            syllables = entry["syllables"]
            display = f"{numeric_attached} {expansion}" if numeric_attached else expansion
            tts = f"{numeric_attached} {syllables}" if numeric_attached else syllables

            normalised_parts.append(display)
            tts_parts.append(tts)
            expansions_applied.append({"from": token, "to": expansion})

            # Only flag when we *defaulted* — i.e. picked one of multiple plausible
            # meanings without a deterministic rule to resolve it. KG resolved by
            # numeric_prefix_is_weight is not a guess; SG defaulting to Sungai is.
            if "ambiguous_with" in entry and entry.get("rule") == "default_to_malay":
                ambiguous_flags.append({
                    "token": token,
                    "chosen": expansion,
                    "alternative": entry["ambiguous_with"],
                    "reason": f"defaulted to {expansion}; could also mean {entry['ambiguous_with']}",
                })
        else:
            # Title-case only purely alphabetic all-caps tokens. Mixed
            # alphanumeric tokens (KT1, A12, 50450) are codes — preserve them.
            pure_alpha = _PURE_ALPHA_RE.match(alpha_part) is not None
            display_alpha = (
                _title_case(alpha_part)
                if pure_alpha and _is_all_caps(alpha_part)
                else alpha_part
            )
            tts = syllabify(display_alpha)
            normalised_parts.append((numeric_attached or "") + display_alpha)
            tts_parts.append((numeric_attached or "") + tts)

        previous_content_token = token

    return {
        "normalised": "".join(normalised_parts),
        "tts_ready": "".join(tts_parts),
        "expansions_applied": expansions_applied,
        "ambiguous_flags": ambiguous_flags,
    }
